File: src/data/campaigns.js
import fs from "fs";
import path from "path";
import crypto from "crypto";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const FILE = path.join(__dirname, "..", "..", "campaigns.json");

// ─────────────────────────────────────────────────────────────
// Outbound campaigns: a named list of {name, phone} recipients
// that the dialer can run in bulk. Persisted to campaigns.json.
// ─────────────────────────────────────────────────────────────

function readAll() {
  try {
    return JSON.parse(fs.readFileSync(FILE, "utf8"));
  } catch {
    return [];
  }
}

function writeAll(rows) {
  fs.writeFileSync(FILE, JSON.stringify(rows, null, 2));
}

export function listCampaigns() {
  return readAll();
}

export function getCampaign(id) {
  return readAll().find((c) => c.id === id) || null;
}

export function createCampaign({ name, recipients = [] }) {
  const rows = readAll();
  const campaign = {
    id: crypto.randomUUID(),
    name: name || "Untitled campaign",
    recipients: recipients
      .map((r) => ({ name: r.name || "", phone: (r.phone || "").trim() }))
      .filter((r) => r.phone),
    status: "draft",
    results: [],
    createdAt: new Date().toISOString(),
  };
  rows.unshift(campaign);
  writeAll(rows);
  return campaign;
}

export function updateCampaign(id, patch) {
  const rows = readAll();
  const idx = rows.findIndex((c) => c.id === id);
  if (idx < 0) return null;
  rows[idx] = { ...rows[idx], ...patch, updatedAt: new Date().toISOString() };
  writeAll(rows);
  return rows[idx];
}

export function deleteCampaign(id) {
  const rows = readAll().filter((c) => c.id !== id);
  writeAll(rows);
}

File: app/data/campaigns.py
"""Outbound campaign storage (ported from src/data/campaigns.js)."""
import json
import os
import uuid
from datetime import datetime, timezone

_ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
_DATA_DIR = os.environ.get("DATA_DIR") or _ROOT
FILE = os.path.join(_DATA_DIR, "campaigns.json")


def _now_iso():
    return datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")


def _read_all():
    try:
        with open(FILE, "r", encoding="utf-8") as f:
            return json.load(f)
    except Exception:
        return []


def _write_all(rows):
    with open(FILE, "w", encoding="utf-8") as f:
        json.dump(rows, f, ensure_ascii=False, indent=2)


def list_campaigns():
    return _read_all()


def get_campaign(cid):
    return next((c for c in _read_all() if c.get("id") == cid), None)


def create_campaign(name=None, recipients=None):
    rows = _read_all()
    recipients = recipients or []
    campaign = {
        "id": str(uuid.uuid4()),
        "name": name or "Untitled campaign",
        "recipients": [
            {"name": r.get("name", ""), "phone": (r.get("phone", "") or "").strip()}
            for r in recipients if (r.get("phone", "") or "").strip()
        ],
        "status": "draft",
        "results": [],
        "createdAt": _now_iso(),
    }
    rows.insert(0, campaign)
    _write_all(rows)
    return campaign


def update_campaign(cid, patch):
    rows = _read_all()
    idx = next((i for i, c in enumerate(rows) if c.get("id") == cid), -1)
    if idx < 0:
        return None
    rows[idx] = {**rows[idx], **patch, "updatedAt": _now_iso()}
    _write_all(rows)
    return rows[idx]


def delete_campaign(cid):
    _write_all([c for c in _read_all() if c.get("id") != cid])
